### PR TITLE
[RPA-1060] Scheduler stopped on bot's registration process

### PR DESCRIPTION
**Pull request by damlit**
[RPA-1060] Scheduler stopped on bot's registration process

## Description

refactor to remove lock on database

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

[RPA-1060]: https://all41.atlassian.net/browse/RPA-1060?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts [hunk 1]
```diff
@@ -35,31 +37,25 @@ export class BotLifecycleService {
...
             ) return;
-            
-            const completeProcessInstance = await this.processInstanceService
-                .findById(disconnectedInstance.id)
-                .catch<null>(error => {
-                    this.logger.error(`Error getting processInstance: ${error}`);
-                    return null;
-                });
-                
+
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/bot-lifecycle.service.ts:40]: lets make this foreach parallel

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.ts [hunk 2]
```diff
@@ -72,6 +72,29 @@ export class BotProcessService {
...
+    }
+
+    async updateProcessLastRunTime(processInstance: IProcessInstance) {
+        if (!processInstance?.process?.id) return;
+        const process = await this.processService.findById(processInstance.process.id);
+
+        if (process && !processInstance.created) {
+            const processInstanceFromDb = await this.processInstanceService.findById(processInstance.id);
+            processInstance.created = processInstanceFromDb.created;
+            await this.processService.save(process);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.ts]: why save process here if no action has been made on this object?

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.ts [hunk 3]
```diff
@@ -72,6 +72,29 @@ export class BotProcessService {
...
+        } else if (process && processInstance.created) {
+            process.lastRun = processInstance.created;
+            await this.processService.save(process);
+        }
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.ts]: imo `if (process)` should be a guard if
then no need to create `else if` - just `else`

Comment (preludium) [reply runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.ts]: why do we need 1st if in the first place? this information should be already in DB and we should not change it

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.ts [hunk 4]
```diff
@@ -72,6 +72,29 @@ export class BotProcessService {
...
         }
     }
+    async updateInterruptedProcessInstance(processInstance: IProcessInstance) {
+        if (!processInstance.rootProcessInstanceId) {
+            this.uiGateway.server.emit(WsMessage.PROCESS, processInstance);
+        }
+        if (isProcessInstanceFinished(processInstance.status)) {
+            await this.processFileService.deleteTempFiles(processInstance.orchestratorProcessInstanceId);
+        }
+        await this.updateProcessLastRunTime(processInstance);
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.ts]: name of this function differs from what it actually does. its handling additional informations/properties about the processInstance

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts [hunk 5]
```diff
@@ -139,6 +108,41 @@ export class BotProcessEventService {
         }
     }
 
+    async updateInterruptedProcessInstanceEvent(
+        queryRunner: QueryRunner, processInstanceEvent: IProcessInstanceEvent, processInstance: IProcessInstance
+    ) {
+        if(!processInstance) return;
+
+        if(getIsInstanceInterrupted(processInstanceEvent.status, processInstance.status)) {
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts]: `getIsInstanceInterrupted` => `isInstanceInterrupted`

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts [hunk 6]
```diff
@@ -139,6 +108,41 @@ export class BotProcessEventService {
         }
     }
 
+    async updateInterruptedProcessInstanceEvent(
+        queryRunner: QueryRunner, processInstanceEvent: IProcessInstanceEvent, processInstance: IProcessInstance
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts:112]: im not sure about passing queryRunner as a parameter

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts [hunk 7]
```diff
@@ -139,6 +108,41 @@ export class BotProcessEventService {
...
+    ) {
+        if(!processInstance) return;
+
+        if(getIsInstanceInterrupted(processInstanceEvent.status, processInstance.status)) {
+            const error = processInstance?.error;
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts]: no need to include `?` you already checked if it exists

**Review by preludium — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts (hunk 6 above)
Comment (damlit) [reply runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance-event.service.ts:112]: It's a part of method which has existed and it's a part of transaction

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.utils.ts [hunk 8]
```diff
@@ -6,7 +6,7 @@ const COMMON_PROCESS_INSTANCE_FIELDS = [
 ];
 
 const COMPLETED_PROCESS_INSTANCE_FIELDS = ['status', 'updated', 'output', ...COMMON_PROCESS_INSTANCE_FIELDS];
-const STARTING_PROCESS_INSTANCE_FIELDS = ['input', 'created', ...COMMON_PROCESS_INSTANCE_FIELDS];
+const STARTING_PROCESS_INSTANCE_FIELDS = ['created', ...COMMON_PROCESS_INSTANCE_FIELDS];
```
Comment (preludium) [inline runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.utils.ts:9]: when is `input` set? when `INITIALIZING` message will arrive?

**Review by preludium — APPROVED** [review]

File: runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.utils.ts (hunk 8 above)
Comment (damlit) [reply runbotics/runbotics-scheduler/src/websocket/bot/process-launch/bot-process-instance.service.utils.ts:9]: Input is set during InProgress, here it was empty

**Pull request merged**